core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -74,6 +74,14 @@ class BlockManagerMasterEndpoint(
   // Mapping from block id to the set of block managers that have the block.
   private val blockLocations = new JHashMap[BlockId, mutable.HashSet[BlockManagerId]]
 
+  // Mapping from host name to shuffle (mergers) services where the current app
+  // registered an executor in the past. Older hosts are removed when the
+  // maxRetainedMergerLocations size is reached in favor of newer locations.
+  private val shuffleMergerLocations = new mutable.LinkedHashMap[String, BlockManagerId]()
```

Comment:
Maybe we could remove the corresponding merger when there's fetch failure happens on it?

Reply:
@Ngone51 That is a good idea, is there any concern with adding this @venkata91 ?
But I would like to add that it would help in only a subset of cases.

What I mean is, when fetch of a merged block fails, executors will fallback to fetching the constituent blocks [1]. A fetch failure would be reported to driver only when both of these fetches fail - the merged block and the mapper output shuffle block fetch. The fallback mechanism would mean that if the individual blocks for a merged block were not computed on the lost host for the parent stage, we wont see the fetch failure.

With the recent changes to prefer hosts with active executors, we do improve the chances of detecting lost merger candidates due to lost hosts though.

[1] Ignoring cases like large blocks which are not candidates for merge, etc.

Reply:
> With the recent changes to prefer hosts with active executors, we do improve the chances of detecting lost merger candidates due to lost hosts though.

Yeah, it shall be fine mostly. But please also note that executors and external shuffle service/merger are two separate processes. So the active executor doesn't necessarily mean the active merger.

I agree it may not help a lot so I'm fine to not add it now.

Reply:
Yes, makes sense. Still I went ahead and made that change. One question I had is what if its a transient failure due to heavy load or something of that sorts? Do we still want to remove the host from the list in those cases?